### PR TITLE
feat: skip logits without liger by default as well

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

In the `SFTTrainer` the logits were materialized for the non liger path, which is not aligned with the liger path. It can result in a huge memory consumption during training.

Context: was using the `SFTTrainer` on Trainium instances and was experiencing OOMs.

---

> [!NOTE]
> **Medium Risk**
> Changes when eval logs entropy and token accuracy on the non-Liger path; setups that relied on those metrics without requesting logits may see different behavior, though `predict()` and metric hooks still force logits when required.
> 
> **Overview**
> **`SFTTrainer.compute_loss`** now computes a shared **`skip_logits`** flag (training, loss-only eval, or eval without `compute_metrics` / `preprocess_logits_for_metrics`, while still honoring **`predict()`** via `_prediction_loss_only`) instead of defining that logic only inside the Liger branch.
> 
> Liger training still sets **`inputs["skip_logits"]`** from that flag. On the **non-Liger** path, entropy and mean token accuracy are only derived from **`outputs.logits`** when **`skip_logits`** is false, so typical training and loss-only eval no longer touch full vocabulary logits for those metrics—reducing memory pressure that showed up as OOMs on large-vocab / Trainium-style runs (see trl#4679).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cabc43501ac09dec44be6abf63d17827f1351f1a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
superseded by https://github.com/huggingface/trl/pull/6314

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1761,7 +1764,7 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             entropy_sum = self.accelerator.gather_for_metrics(outputs.entropy_sum).sum()
             entropy = (entropy_sum / num_valid).item() if num_valid > 0 else 0.0
             self._metrics[mode]["entropy"].append(entropy)
-        elif not self.args.use_liger_kernel:  # liger doesn't return logits
+        elif not self.args.use_liger_kernel and not skip_logits:  # liger doesn't return logits
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:1766]: Yes, I think this is a real problem. This PR would effectively drop the `entropy` and `mean_token_accuracy` metrics (two important metrics) for all non-Liger trainings with `loss_type="nll"`, since `skip_logits` is always `True` while `self.model.training is True`.

That said, I'm not convinced this PR is actually needed for SFT on Trainium. You can simply keep the default `loss_type="chunked_nll"`, which already skips the `lm_head` projection on ignored tokens and avoids the OOM without dropping any metrics.

Is there a case where `chunked_nll` isn't suitable and `nll` is required? The only one I can think of is having `lm_head` wrapped by a PEFT adapter (`target_modules="all-linear"`). Is this your actual setup?

[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply trl/trainer/sft_trainer.py:1766]: I agree, it is not required for Neuron. The question then is: do we want a different behavior when we use the liger kernels (no metrics) compared to when we dont use it?

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:1766]: liger outputs a couple of metrics, but you're right we've no guaranty that they match the one we've in trl. Usually we try to contribute liger to align. But I think this is not a big deal, this is already the case in sft, dpo etc

**Comment by michaelbenayoun** [comment]
Alright, closing the PR!
